File: packages/ios/Sources/OnloSDK/MobileConfig.swift
import Foundation
import Security

/// The complete, supported v1 mobile configuration.  All documented fields are
/// required; Codable deliberately ignores additive fields the server may add.
public struct MobileConfig: Codable, Sendable, Equatable {
    public let schemaVersion: Int
    public let revision: String
    public let compatibility: Compatibility
    public let securityPolicy: SecurityPolicy
    public let appearance: Appearance
    public let features: Features
    public let mediaPolicy: MediaPolicy
    public let content: Content
    public let identityMode: IdentityMode
    public let unsupportedWidgetSettings: [UnsupportedWidgetSetting]

    public init(from decoder: Decoder) throws {
        let value = try Self.decodeKnownFields(from: decoder)
        guard value.schemaVersion == 1,
              value.compatibility.requestedSchemaVersion == 1,
              value.compatibility.appliedSchemaVersion == 1,
              value.securityPolicy.minimumProtocolVersion == OnloProtocol.version,
              value.securityPolicy.identityMode == .sdkInterface,
              value.securityPolicy.anonymousScope == .installationGeneration,
              value.securityPolicy.nativePlacement == .hostApp,
              value.identityMode == .sdkInterface else {
            throw OnloError.incompatibleProtocol
        }
        self = value
    }

    private static func decodeKnownFields(from decoder: Decoder) throws -> MobileConfig {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        return try MobileConfig(
            schemaVersion: c.decode(Int.self, forKey: .schemaVersion),
            revision: c.decode(String.self, forKey: .revision),
            compatibility: c.decode(Compatibility.self, forKey: .compatibility),
            securityPolicy: c.decode(SecurityPolicy.self, forKey: .securityPolicy),
            appearance: c.decode(Appearance.self, forKey: .appearance),
            features: c.decode(Features.self, forKey: .features),
            mediaPolicy: c.decode(MediaPolicy.self, forKey: .mediaPolicy),
            content: c.decode(Content.self, forKey: .content),
            identityMode: c.decode(IdentityMode.self, forKey: .identityMode),
            unsupportedWidgetSettings: c.decode([UnsupportedWidgetSetting].self, forKey: .unsupportedWidgetSettings)
        )
    }

    private init(schemaVersion: Int, revision: String, compatibility: Compatibility, securityPolicy: SecurityPolicy, appearance: Appearance, features: Features, mediaPolicy: MediaPolicy, content: Content, identityMode: IdentityMode, unsupportedWidgetSettings: [UnsupportedWidgetSetting]) {
        self.schemaVersion = schemaVersion; self.revision = revision; self.compatibility = compatibility; self.securityPolicy = securityPolicy
        self.appearance = appearance; self.features = features; self.mediaPolicy = mediaPolicy; self.content = content; self.identityMode = identityMode; self.unsupportedWidgetSettings = unsupportedWidgetSettings
    }

    public struct Compatibility: Codable, Sendable, Equatable { public let requestedSchemaVersion: Int; public let appliedSchemaVersion: Int; public let capabilities: [MobileCapability]; public let unsupportedSettings: [UnsupportedSetting] }
    public struct UnsupportedSetting: Codable, Sendable, Equatable { public let code: String; public let setting: String; public let reason: String; public let requiredCapabilities: [MobileCapability]? }
    public struct SecurityPolicy: Codable, Sendable, Equatable {
        public let minimumProtocolVersion: Int; public let minimumSdkVersion: String?; public let identityMode: IdentityMode; public let anonymousScope: AnonymousScope; public let nativePlacement: NativePlacement
        private enum CodingKeys: String, CodingKey { case minimumProtocolVersion, minimumSdkVersion, identityMode, anonymousScope, nativePlacement }
        public init(from decoder: Decoder) throws { let c = try decoder.container(keyedBy: CodingKeys.self); minimumProtocolVersion = try c.decode(Int.self, forKey: .minimumProtocolVersion); minimumSdkVersion = try c.decodeIfPresent(String.self, forKey: .minimumSdkVersion); identityMode = try c.decode(IdentityMode.self, forKey: .identityMode); anonymousScope = try c.decode(AnonymousScope.self, forKey: .anonymousScope); nativePlacement = try c.decode(NativePlacement.self, forKey: .nativePlacement) }
    }
    public struct Appearance: Codable, Sendable, Equatable { public let accent: String; public let botName: String; public let botSubtitle: String; public let greeting: String; public let headerAvatar: HeaderAvatar; public let light: ColorTheme; public let dark: DarkColorTheme }
    public struct HeaderAvatar: Codable, Sendable, Equatable {
        public let mode: AvatarMode; public let text: String; public let data: String?
        private enum CodingKeys: String, CodingKey { case mode, text, data }
        public init(from decoder: Decoder) throws { let c = try decoder.container(keyedBy: CodingKeys.self); mode = try c.decode(AvatarMode.self, forKey: .mode); text = try c.decode(String.self, forKey: .text); data = try c.decodeIfPresent(String.self, forKey: .data) }
    }
    public struct ColorTheme: Codable, Sendable, Equatable { public let background: String; public let outgoing: String; public let outgoingText: String; public let incoming: String; public let incomingText: String }
    public struct DarkColorTheme: Codable, Sendable, Equatable { public let enabled: Bool; public let background: String; public let outgoing: String; public let outgoingText: String; public let incoming: String; public let incomingText: String }
    public struct Features: Codable, Sendable, Equatable { public let insertLink: Bool; public let insertCode: Bool; public let emoji: Bool; public let gifs: Bool; public let voice: Bool; public let fileUpload: Bool; public let transcriptDownload: Bool; public let soundNotifications: Bool; public let showTimestamps: Bool; public let faqButton: FAQButton }
    public struct MediaPolicy: Codable, Sendable, Equatable {
        public let enabled: Bool
        public let maximumImagesPerMessage: Int
        public let maximumImageBytes: Int

        private enum CodingKeys: String, CodingKey {
            case enabled, maximumImagesPerMessage, maximumImageBytes
        }

        public init(from decoder: Decoder) throws {
            let c = try decoder.container(keyedBy: CodingKeys.self)
            enabled = try c.decode(Bool.self, forKey: .enabled)
            maximumImagesPerMessage = try c.decode(Int.self, forKey: .maximumImagesPerMessage)
            maximumImageBytes = try c.decode(Int.self, forKey: .maximumImageBytes)
            guard (0...OnloProtocol.maximumImagesPerMessage).contains(maximumImagesPerMessage),
                  (1...OnloProtocol.maximumImageBytes).contains(maximumImageBytes) else {
                throw OnloError.invalidResponse
            }
        }

        public var effectiveMaximumImagesPerMessage: Int {
            min(maximumImagesPerMessage, OnloProtocol.maximumImagesPerMessage)
        }

        public var effectiveMaximumImageBytes: Int {
            min(maximumImageBytes, OnloProtocol.maximumImageBytes)
        }
    }
    public struct FAQButton: Codable, Sendable, Equatable { public let enabled: Bool; public let label: String }
    public struct Content: Codable, Sendable, Equatable { public let faqs: [FAQ]; public let tabs: Tabs; public let search: Search; public let onboarding: Onboarding; public let homeSections: [HomeSection] }
    public struct FAQ: Codable, Sendable, Equatable { public let question: String; public let answer: String? }
    public struct Tabs: Codable, Sendable, Equatable { public let enabled: Bool; public let tabs: [Tab]; public let defaultTab: String }
    public struct Tab: Codable, Sendable, Equatable { public let id: String; public let label: String; public let icon: String; public let enabled: Bool }
    public struct Search: Codable, Sendable, Equatable { public let enabled: Bool; public let placeholder: String; public let showSearchInHome: Bool }
    public struct Onboarding: Codable, Sendable, Equatable { public let enabled: Bool; public let title: String; public let showProgress: Bool; public let items: [OnboardingItem] }
    public struct OnboardingItem: Codable, Sendable, Equatable { public let id: String; public let title: String; public let description: String?; public let completed: Bool; public let actionUrl: String? }
    public struct HomeSection: Codable, Sendable, Equatable { public let id: String; public let type: HomeSectionType; public let title: String?; public let content: String?; public let enabled: Bool; public let order: Int }
    public struct UnsupportedWidgetSetting: Codable, Sendable, Equatable { public let setting: String; public let reason: String }
}

public enum MobileCapability: String, Codable, Sendable, Equatable { case secureStorage = "secure_storage", persistentOutbox = "persistent_outbox", foregroundStream = "foreground_stream", apns, fcm, mediaPicker = "media_picker", attachmentUpload = "attachment_upload", configSchemaV1 = "config_schema_v1", identityJWT = "identity_jwt", appAttestation = "app_attestation", deepLinkRouting = "deep_link_routing" }
public enum IdentityMode: String, Codable, Sendable, Equatable { case sdkInterface = "sdk_interface" }
public enum AnonymousScope: String, Codable, Sendable, Equatable { case installationGeneration = "installation_generation" }
public enum NativePlacement: String, Codable, Sendable, Equatable { case hostApp = "host_app" }
public enum AvatarMode: String, Codable, Sendable, Equatable { case image, initials }
public enum HomeSectionType: String, Codable, Sendable, Equatable { case welcome, search, faqs, checklist, custom }

struct ProtectedMobileConfigState: Codable, Sendable, Equatable {
    let config: MobileConfig?
    let etag: String?
    let retry: ConfigRetryState?
}

extension ProtectedMobileConfigState {
    func validated() throws -> ProtectedMobileConfigState {
        // A retry-only state has neither value. Every cached configuration must
        // have the validator that authorises its conditional reuse, and vice versa.
        guard (config == nil) == (etag == nil) else { throw OnloError.credentialStore(code: "config_state_invariant_failed") }
        return self
    }
}

enum ConfigRetryState: Codable, Sendable, Equatable { case afterBackoff(eligibleAt: Date, attempt: Int) }

protocol MobileConfigStoring: Sendable {
    func loadConfigState() async throws -> ProtectedMobileConfigState
    func saveConfigState(_ state: ProtectedMobileConfigState) async throws
}

protocol AuthorityFencedConfigStoring: Sendable {
    func activateAuthority(_ authority: PersistenceAuthority) async
    func revokeAuthority(for scope: OwnerScope) async
    func saveConfigState(
        _ state: ProtectedMobileConfigState,
        authority: PersistenceAuthority
    ) async throws -> Bool
}

/// A single Keychain record makes config and ETag updates atomic and prevents a
/// process observing a new ETag beside an older configuration. Keychain is the
/// protected/encrypted persistence boundary; no configuration is stored in
/// UserDefaults or an ordinary file.
actor KeychainMobileConfigStore: MobileConfigStoring, AuthorityFencedConfigStoring {
    private let service = "ai.onlo.sdk.mobile-config"
    private let account = "v1"
    private let encoder = JSONEncoder()
    private let decoder = JSONDecoder()
    private var activeAuthority: PersistenceAuthority?

    func loadConfigState() async throws -> ProtectedMobileConfigState {
        let query: [String: Any] = [kSecClass as String: kSecClassGenericPassword, kSecAttrService as String: service, kSecAttrAccount as String: account, kSecReturnData as String: true, kSecMatchLimit as String: kSecMatchLimitOne]
        var result: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &result)
        if status == errSecItemNotFound { return ProtectedMobileConfigState(config: nil, etag: nil, retry: nil) }
        guard status == errSecSuccess, let data = result as? Data else { throw OnloError.credentialStore(code: "config_keychain_read_failed") }
        do { return try decoder.decode(ProtectedMobileConfigState.self, from: data).validated() }
        catch { throw OnloError.credentialStore(code: "config_keychain_decode_failed") }
    }

    func saveConfigState(_ state: ProtectedMobileConfigState) async throws {
        try saveConfigStateSynchronously(state)
    }

    private func saveConfigStateSynchronously(_ state: ProtectedMobileConfigState) throws {
        _ = try state.validated()
        let data: Data
        do { data = try encoder.encode(state) } catch { throw OnloError.credentialStore(code: "config_keychain_encode_failed") }
        let query: [String: Any] = [kSecClass as String: kSecClassGenericPassword, kSecAttrService as String: service, kSecAttrAccount as String: account]
        let attributes: [String: Any] = [kSecValueData as String: data, kSecAttrAccessible as String: kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly]
        let update = SecItemUpdate(query as CFDictionary, attributes as CFDictionary)
        if update == errSecSuccess { return }
        guard update == errSecItemNotFound else { throw OnloError.credentialStore(code: "config_keychain_write_failed") }
        var add = query; add.merge(attributes) { _, new in new }
        guard SecItemAdd(add as CFDictionary, nil) == errSecSuccess else { throw OnloError.credentialStore(code: "config_keychain_write_failed") }
    }

    func activateAuthority(_ authority: PersistenceAuthority) {
        activeAuthority = authority
    }

    func revokeAuthority(for scope: OwnerScope) {
        if activeAuthority?.ownerScope == scope { activeAuthority = nil }
    }

    func saveConfigState(
        _ state: ProtectedMobileConfigState,
        authority: PersistenceAuthority
    ) async throws -> Bool {
        guard activeAuthority == authority else { return false }
        try saveConfigStateSynchronously(state)
        return true
    }
}

File: packages/android/src/main/kotlin/ai/onlo/sdk/config/MobileConfig.kt
package ai.onlo.sdk.config

import ai.onlo.sdk.protocol.Capability
import ai.onlo.sdk.protocol.ProtocolViolation
import ai.onlo.sdk.protocol.requiredBoolean
import ai.onlo.sdk.protocol.requiredInt
import ai.onlo.sdk.protocol.requiredObject
import ai.onlo.sdk.protocol.requiredString
import org.json.JSONArray
import org.json.JSONObject

/** Strict, read-only representation of the server-owned mobile configuration schema v1. */
public data class MobileConfig(
    val schemaVersion: Int,
    val revision: String,
    val compatibility: Compatibility,
    val securityPolicy: SecurityPolicy,
    val appearance: Appearance,
    val features: Features,
    val mediaPolicy: MediaPolicy,
    val content: Content,
    val identityMode: String,
    val unsupportedWidgetSettings: List<UnsupportedWidgetSetting>,
) {
    public data class Compatibility(val requestedSchemaVersion: Int, val appliedSchemaVersion: Int, val capabilities: List<Capability>, val unsupportedSettings: List<UnsupportedSetting>)
    public data class UnsupportedSetting(val code: String, val setting: String, val reason: String, val requiredCapabilities: List<Capability>?)
    public data class SecurityPolicy(val minimumProtocolVersion: Int, val minimumSdkVersion: String?, val identityMode: String, val anonymousScope: String, val nativePlacement: String)
    public data class Appearance(val accent: String, val botName: String, val botSubtitle: String, val greeting: String, val headerAvatar: HeaderAvatar, val light: ColorTheme, val dark: DarkColorTheme)
    public data class HeaderAvatar(val mode: HeaderAvatarMode, val text: String, val data: String?)
    public enum class HeaderAvatarMode { IMAGE, INITIALS }
    public data class ColorTheme(val background: String, val outgoing: String, val outgoingText: String, val incoming: String, val incomingText: String)
    /** This maps the contract's flat `ColorTheme & { enabled }` shape exactly. */
    public data class DarkColorTheme(val enabled: Boolean, val background: String, val outgoing: String, val outgoingText: String, val incoming: String, val incomingText: String)
    public data class Features(val insertLink: Boolean, val insertCode: Boolean, val emoji: Boolean, val gifs: Boolean, val voice: Boolean, val fileUpload: Boolean, val transcriptDownload: Boolean, val soundNotifications: Boolean, val showTimestamps: Boolean, val faqButton: FaqButton)
    public data class MediaPolicy(val enabled: Boolean, val maximumImagesPerMessage: Int, val maximumImageBytes: Int) {
        public val effectiveMaximumImagesPerMessage: Int get() = maximumImagesPerMessage.coerceAtMost(5)
        public val effectiveMaximumImageBytes: Int get() = maximumImageBytes.coerceAtMost(8_388_608)
    }
    public data class FaqButton(val enabled: Boolean, val label: String)
    public data class Content(val faqs: List<Faq>, val tabs: Tabs, val search: Search, val onboarding: Onboarding, val homeSections: List<HomeSection>)
    public data class Faq(val question: String, val answer: String?)
    public data class Tabs(val enabled: Boolean, val tabs: List<Tab>, val defaultTab: String)
    public data class Tab(val id: String, val label: String, val icon: String, val enabled: Boolean)
    public data class Search(val enabled: Boolean, val placeholder: String, val showSearchInHome: Boolean)
    public data class Onboarding(val enabled: Boolean, val title: String, val showProgress: Boolean, val items: List<OnboardingItem>)
    public data class OnboardingItem(val id: String, val title: String, val description: String?, val completed: Boolean, val actionUrl: String?)
    public data class HomeSection(val id: String, val type: HomeSectionType, val title: String?, val content: String?, val enabled: Boolean, val order: Int)
    public enum class HomeSectionType { WELCOME, SEARCH, FAQS, CHECKLIST, CUSTOM }
    public data class UnsupportedWidgetSetting(val setting: String, val reason: String)
}

/** The raw JSON is retained only in encrypted native storage after this decoder accepts it. */
internal object MobileConfigCodec {
    fun decode(raw: String): MobileConfig = decode(JSONObject(raw))

    fun decode(value: JSONObject): MobileConfig {
        val config = MobileConfig(
            schemaVersion = value.requiredInt("schemaVersion").also { requireProtocol(it == 1, "config_schema") },
            revision = value.requiredString("revision"),
            compatibility = decodeCompatibility(value.requiredObject("compatibility")),
            securityPolicy = decodeSecurity(value.requiredObject("securityPolicy")),
            appearance = decodeAppearance(value.requiredObject("appearance")),
            features = decodeFeatures(value.requiredObject("features")),
            mediaPolicy = decodeMediaPolicy(value.requiredObject("mediaPolicy")),
            content = decodeContent(value.requiredObject("content")),
            identityMode = value.requiredString("identityMode"),
            unsupportedWidgetSettings = array(value, "unsupportedWidgetSettings").map { item ->
                MobileConfig.UnsupportedWidgetSetting(item.requiredString("setting"), item.requiredString("reason"))
            },
        )
        requireProtocol(config.compatibility.requestedSchemaVersion == 1 && config.compatibility.appliedSchemaVersion == 1, "config_compatibility")
        requireProtocol(config.securityPolicy.minimumProtocolVersion == 1, "config_protocol")
        requireProtocol(config.securityPolicy.identityMode == "sdk_interface" && config.identityMode == "sdk_interface", "config_identity_mode")
        requireProtocol(config.securityPolicy.anonymousScope == "installation_generation", "config_anonymous_scope")
        requireProtocol(config.securityPolicy.nativePlacement == "host_app", "config_native_placement")
        return config
    }

    private fun decodeCompatibility(v: JSONObject) = MobileConfig.Compatibility(
        v.requiredInt("requestedSchemaVersion"), v.requiredInt("appliedSchemaVersion"),
        stringArray(v, "capabilities").map(::capability),
        array(v, "unsupportedSettings").map { item -> MobileConfig.UnsupportedSetting(item.requiredString("code"), item.requiredString("setting"), item.requiredString("reason"), optionalStringArray(item, "requiredCapabilities")?.map(::capability)) },
    )
    private fun decodeSecurity(v: JSONObject) = MobileConfig.SecurityPolicy(v.requiredInt("minimumProtocolVersion"), requiredNullableString(v, "minimumSdkVersion"), v.requiredString("identityMode"), v.requiredString("anonymousScope"), v.requiredString("nativePlacement"))
    private fun decodeAppearance(v: JSONObject): MobileConfig.Appearance {
        val avatar = v.requiredObject("headerAvatar")
        val mode = when (avatar.requiredString("mode")) { "image" -> MobileConfig.HeaderAvatarMode.IMAGE; "initials" -> MobileConfig.HeaderAvatarMode.INITIALS; else -> throw ProtocolViolation("header_avatar_mode") }
        val dark = v.requiredObject("dark")
        return MobileConfig.Appearance(v.requiredString("accent"), v.requiredString("botName"), v.requiredString("botSubtitle"), v.requiredString("greeting"), MobileConfig.HeaderAvatar(mode, avatar.requiredString("text"), requiredNullableString(avatar, "data")), colors(v.requiredObject("light")), MobileConfig.DarkColorTheme(dark.requiredBoolean("enabled"), dark.requiredString("background"), dark.requiredString("outgoing"), dark.requiredString("outgoingText"), dark.requiredString("incoming"), dark.requiredString("incomingText")))
    }
    private fun colors(v: JSONObject) = MobileConfig.ColorTheme(v.requiredString("background"), v.requiredString("outgoing"), v.requiredString("outgoingText"), v.requiredString("incoming"), v.requiredString("incomingText"))
    private fun decodeFeatures(v: JSONObject): MobileConfig.Features {
        val faq = v.requiredObject("faqButton")
        return MobileConfig.Features(v.requiredBoolean("insertLink"), v.requiredBoolean("insertCode"), v.requiredBoolean("emoji"), v.requiredBoolean("gifs"), v.requiredBoolean("voice"), v.requiredBoolean("fileUpload"), v.requiredBoolean("transcriptDownload"), v.requiredBoolean("soundNotifications"), v.requiredBoolean("showTimestamps"), MobileConfig.FaqButton(faq.requiredBoolean("enabled"), faq.requiredString("label")))
    }
    private fun decodeMediaPolicy(v: JSONObject): MobileConfig.MediaPolicy {
        val maximumImagesPerMessage = v.requiredInt("maximumImagesPerMessage")
        val maximumImageBytes = v.requiredInt("maximumImageBytes")
        requireProtocol(maximumImagesPerMessage in 0..5, "config_media_maximum_images")
        requireProtocol(maximumImageBytes in 1..8_388_608, "config_media_maximum_bytes")
        return MobileConfig.MediaPolicy(v.requiredBoolean("enabled"), maximumImagesPerMessage, maximumImageBytes)
    }
    private fun decodeContent(v: JSONObject) = MobileConfig.Content(
        array(v, "faqs").map { MobileConfig.Faq(it.requiredString("question"), optionalString(it, "answer")) },
        decodeTabs(v.requiredObject("tabs")),
        v.requiredObject("search").let { MobileConfig.Search(it.requiredBoolean("enabled"), it.requiredString("placeholder"), it.requiredBoolean("showSearchInHome")) },
        decodeOnboarding(v.requiredObject("onboarding")),
        array(v, "homeSections").map { item -> MobileConfig.HomeSection(item.requiredString("id"), homeType(item.requiredString("type")), optionalString(item, "title"), optionalString(item, "content"), item.requiredBoolean("enabled"), item.requiredInt("order")) },
    )
    private fun decodeTabs(v: JSONObject) = MobileConfig.Tabs(v.requiredBoolean("enabled"), array(v, "tabs").map { MobileConfig.Tab(it.requiredString("id"), it.requiredString("label"), it.requiredString("icon"), it.requiredBoolean("enabled")) }, v.requiredString("defaultTab"))
    private fun decodeOnboarding(v: JSONObject) = MobileConfig.Onboarding(v.requiredBoolean("enabled"), v.requiredString("title"), v.requiredBoolean("showProgress"), array(v, "items").map { MobileConfig.OnboardingItem(it.requiredString("id"), it.requiredString("title"), optionalString(it, "description"), it.requiredBoolean("completed"), optionalString(it, "actionUrl")) })
    private fun homeType(value: String): MobileConfig.HomeSectionType = when (value) { "welcome" -> MobileConfig.HomeSectionType.WELCOME; "search" -> MobileConfig.HomeSectionType.SEARCH; "faqs" -> MobileConfig.HomeSectionType.FAQS; "checklist" -> MobileConfig.HomeSectionType.CHECKLIST; "custom" -> MobileConfig.HomeSectionType.CUSTOM; else -> throw ProtocolViolation("home_section_type") }
    private fun capability(value: String): Capability = Capability.entries.firstOrNull { it.wireValue == value } ?: throw ProtocolViolation("config_capability")
    private fun array(v: JSONObject, name: String): List<JSONObject> = try { (v.get(name) as? JSONArray ?: throw ProtocolViolation(name)).let { values -> List(values.length()) { index -> values.get(index) as? JSONObject ?: throw ProtocolViolation(name) } } } catch (e: ProtocolViolation) { throw e } catch (_: Exception) { throw ProtocolViolation(name) }
    private fun stringArray(v: JSONObject, name: String): List<String> = try { (v.get(name) as? JSONArray ?: throw ProtocolViolation(name)).let { values -> List(values.length()) { index -> values.get(index) as? String ?: throw ProtocolViolation(name) } } } catch (e: ProtocolViolation) { throw e } catch (_: Exception) { throw ProtocolViolation(name) }
    private fun optionalStringArray(v: JSONObject, name: String): List<String>? = if (!v.has(name) || v.isNull(name)) null else stringArray(v, name)
    private fun optionalString(v: JSONObject, name: String): String? = if (!v.has(name) || v.isNull(name)) null else v.requiredString(name)
    private fun requiredNullableString(v: JSONObject, name: String): String? { if (!v.has(name)) throw ProtocolViolation(name); return if (v.isNull(name)) null else v.requiredString(name) }
    private fun requireProtocol(condition: Boolean, code: String) { if (!condition) throw ProtocolViolation(code) }
}
